dipy/reconst/ivim.py

```diff
@@ -0,0 +1,567 @@
+            Use the Jacobian. If true, use the Jacobian function defined in
+            `_ivim_jacobian_func` to calculate the Jacobian for minimization.
+
+        bounds : tuple, optional
+            Bounds for the various parameters,
+
+        tol : float, optional
+            Default : 1e-25
+            Tolerance for convergence of minimization
+
+        Returns
+        -------
+        IvimFit object
+
+        References
+        ----------
+        .. [1] Federau, Christian, et al. "Quantitative measurement
+                   of brain perfusion with intravoxel incoherent motion
+                   MR imaging." Radiology 265.3 (2012): 874-881.
+        """
+        if mask is None:
+            # Flatten it to 2D either way:
+            data_in_mask = np.reshape(data, (-1, data.shape[-1]))
+        else:
+            # Check for valid shape of the mask
+            if mask.shape != data.shape[:-1]:
+                raise ValueError("Mask is not the same shape as data.")
+            mask = np.array(mask, dtype=bool, copy=False)
+            data_in_mask = np.reshape(data[mask], (-1, data.shape[-1]))
+        if self.min_signal is None:
+            min_signal = _min_positive_signal(data)
+        else:
+            min_signal = self.min_signal
+
+        if len(self.x0) == 4:
+            x0 = np.ones(
+                (data.shape[:-1] + (4,))) * self.x0
+        else:
+            if self.x0.shape != (data.shape[0], 4):
+                raise ValueError(
+                    "Initial guess params should be of shape (num_voxels,4)")
+
+        data_in_mask = np.maximum(data_in_mask, min_signal)
+        # Get the x0 parameters for the masked data
+        x0_mask = np.reshape(x0[mask], (-1, x0.shape[-1]))
+
+        if self.fit_method == "one_stage":
+            params_in_mask, fit_stats_in_mask = one_stage(data_in_mask,
+                                                          self.gtab, x0_mask,
+                                                          self.jac, self.bounds, self.tol,
+                                                          self.options)
+        elif self.fit_method == "two_stage":
+            params_in_mask, fit_stats_in_mask = two_stage(data_in_mask,
+                                                          self.gtab, x0_mask,
+                                                          self.split_b, self.jac,
+                                                          self.bounds, self.tol,
+                                                          self.options)
+
+        if mask is None:
+            out_shape = data.shape[:-1] + (-1, )
+            ivim_params = params_in_mask.reshape(out_shape)
+            fit_stats = fit_stats_in_mask.reshape(out_shape)
+        else:
+            ivim_params = np.zeros(data.shape[:-1] + (4,))
+            ivim_params[mask, :] = params_in_mask
+            fit_stats = np.zeros(data.shape[:-1] + (4,))
+            fit_stats[mask, :] = fit_stats_in_mask
```

Comment:
Setting the mask is similar to slicing out a part of the data right ? So instead of doing this
```
x1, x2 = 160, 180
y1, y2 = 90, 110
z1, z2 = 0,10
data_slice = data[x1:x2, y1:y2, z1:z2, :]
```
I can set an appropiate mask. I will test this tomorrow.